bundles/automation/org.eclipse.smarthome.automation.core/src/main/java/org/eclipse/smarthome/automation/core/RuleEngine.java

```diff
@@ -0,0 +1,426 @@
+ * http://www.prosyst.com
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *    ProSyst Software GmbH - initial API and implementation
+ *******************************************************************************/
+
+package org.eclipse.smarthome.automation.core;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+import org.eclipse.smarthome.automation.Action;
+import org.eclipse.smarthome.automation.Condition;
+import org.eclipse.smarthome.automation.Connection;
+import org.eclipse.smarthome.automation.Module;
+import org.eclipse.smarthome.automation.Rule;
+import org.eclipse.smarthome.automation.Trigger;
+import org.eclipse.smarthome.automation.core.RuleEngineCallbackImpl.TriggerData;
+import org.eclipse.smarthome.automation.handler.ActionHandler;
+import org.eclipse.smarthome.automation.handler.ConditionHandler;
+import org.eclipse.smarthome.automation.handler.ModuleHandler;
+import org.eclipse.smarthome.automation.handler.ModuleHandlerFactory;
+import org.eclipse.smarthome.automation.handler.RuleEngineCallback;
+import org.eclipse.smarthome.automation.handler.TriggerHandler;
+import org.eclipse.smarthome.automation.type.Input;
+import org.osgi.framework.BundleContext;
+import org.osgi.framework.ServiceReference;
+import org.osgi.util.tracker.ServiceTracker;
+import org.osgi.util.tracker.ServiceTrackerCustomizer;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * @author Yordan Mihaylov - Initial Contribution
+ */
+public class RuleEngine implements ServiceTrackerCustomizer/* <ModuleHandlerFactory, ModuleHandlerFactory> */{
+
+    public static final int MODULE_TYPE_SEPARATOR = ':';
+
+    private static Map<String, RuleImpl> rules;
+    private static Map<String, Thread> runningRules;
+    private ServiceTracker/* <ModuleHandlerFactory, ModuleHandlerFactory> */msfTracker;
+    private BundleContext bc;
+    private Map<String, ModuleHandlerFactory> moduleHandlerFactories;
+    private Logger log;
+
+    public RuleEngine(BundleContext bc) {
+        this.bc = bc;
+        this.log = LoggerFactory.getLogger(getClass());
+        if (rules == null) {
+            rules = new HashMap<String, RuleImpl>(20);
+        }
+        if (runningRules == null) {
+            runningRules = new HashMap<String, Thread>(20);
+        }
+        moduleHandlerFactories = new HashMap<String, ModuleHandlerFactory>(20);
+        msfTracker = new ServiceTracker/* <ModuleHandlerFactory, ModuleHandlerFactory> */(bc,
+                ModuleHandlerFactory.class.getName(), this);
+        msfTracker.open();
+
+    }
+
+    protected void setRule(RuleImpl r) {
```

Comment:
Hm, I am wondering, if we should define some lifecycle for the rule status and have clear statuses instead of "running", "initialized" and "enabled" flags with errorCodes that only make sense in certain states.
It would be nice, if it could be streamlined with the Thing status (https://www.eclipse.org/smarthome/documentation/concepts/things.html#thing-status). There, we have a status object, which has a main status (for rules, this could be NOT_INITIALIZED, IDLE, RUNNING) and a status details (which could be here HANDLER_MISSING_ERROR etc.) and a text message for logging purposes.

I think the requirements are pretty similar and thus the classes could be created in the same way as https://github.com/eclipse/smarthome/blob/8078ab2d0f8da486563dc03df71e87efa7e8a277/bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/ThingStatusInfo.java.